### ISSUE TITLE
[no-inferrable-types] Feature request

### ISSUE THREAD

**Issue by ilyub** [typescript-eslint/typescript-eslint#3722]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-inferrable-types": "warn"
  }
}
```

```TS
function f<T extends object>(...): T {
  const result: T = {} as T;
  ...
  return result;
}
```

**Expected Result**

Feature request: The rule may suggest to remove either ": T" or "as T".

**Actual Result**

I know of no rule that covers this.

**Additional Info**

No

**Versions**

Irrelevant. This is not a bug report.

**Comment by bradzacher** [comment]
> The rule may suggest to remove either ": T" or "as T".

The rule cannot suggest removing `as T` because type assertions are looser than type annotations.
`as T` could be coercing the type in a way that `: T` cannot, thus removing it would be an error.

**Comment by ilyub** [comment]
I agree that removing "as T" is not always possible, but it is sometimes possible. So, the rule can either auto-remove ": T" or it can suggest to the user to choose manually what is possible in each case.

Thx for putting my suggestion in a queue for enhancement.

**Comment by JoshuaKGoldberg** [comment]
Now that ESLint rules can provide _suggestions_ (`hasSuggestions`) separate from _fixes_ (`fixable`), I think we can take this in as a rule suggestion.

**Comment by yeonjuan** [comment]
> no-inferrable-types
> Disallow explicit type declarations for variables or parameters initialized to a number, string, or boolean.

Currently, this rule only works for a few types, such as boolean and string...
[playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=MYGwhgzhAEBiD29oG8BQ1oAcBO9MC5oIAXbASwDsBzAblQF9VVh4KToAzRQhJAXmgUApgHc4iABQBKGkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDNFpoAQwBGSWkVKoMkftA7RI4MAF8QyoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
```ts
class Foo {
  prop: string;
}
const foo: Foo = new Foo(); // no error
```

Does this issue mean adding functionality to check for types other than the ones we currently check for?
If not, does it mean giving “suggestions” for the type you are currently checking?

```ts
const a: boolean = true as boolean;
```

**Comment by JoshuaKGoldberg** [comment]
> Does this issue mean adding functionality to check for types other than the ones we currently check for?

No, though that would be lovely as a separate change. That change would be a breaking one as it would require using the type checker. Which... might be worth it now that we have an assignability API? TBD, investigation needed.

**Comment by JoshuaKGoldberg** [comment]
The rule actually does fix when it can!

https://github.com/typescript-eslint/typescript-eslint/blob/f22de044180498ea4bde3874f9e0a4e6549adcde/packages/eslint-plugin/src/rules/no-inferrable-types.ts#L226-L240

This specific issue for adding a _fixer_ is resolved. ✅
